src/lib/Handshake.js

```diff
@@ -194,79 +202,80 @@ class Handshake {
-      return '';
+    if (decryptedHandshakeData.length < (NONCE_BYTES + ID_BYTES)) {
+      throw new Error(`handshake data was too small: ${decryptedHandshakeData.length}`);
     }
 
-    //success
-    const nonceBuffer = new Buffer(40);
-    const deviceIDBuffer = new Buffer(12);
 
-    plaintext.copy(nonceBuffer, 0, 0, 40);
-    plaintext.copy(deviceIDBuffer, 0, 40, 52);
+    const nonceBuffer = new Buffer(NONCE_BYTES);
+    const deviceIDBuffer = new Buffer(ID_BYTES);
+    const deviceKeyBuffer = new Buffer(
+      decryptedHandshakeData.length - (NONCE_BYTES + ID_BYTES),
+    );
 
-		const deviceKey = new Buffer(plaintext.length - 52);
-		plaintext.copy(deviceKey, 0, 52, plaintext.length);
-		const deviceProvidedPem = utilities.convertDERtoPEM(deviceKey);
+    decryptedHandshakeData.copy(nonceBuffer, 0, 0, NONCE_BYTES);
+    decryptedHandshakeData.copy(
+      deviceIDBuffer,
+      0,
+      NONCE_BYTES,
+      (NONCE_BYTES + ID_BYTES)
+    );
+    decryptedHandshakeData.copy(
+      deviceKeyBuffer,
+      0,
+      (NONCE_BYTES + ID_BYTES),
+      decryptedHandshakeData.length
+    );
 
-    //nonces should match
     if (!utilities.bufferCompare(nonceBuffer, nonce)) {
-      this._handshakeFail('nonces didn\'t match');
-      return '';
+      throw new Error('nonces didn\`t match');
     }
 
-    this._deviceID = deviceIDBuffer.toString('hex');
+		const deviceProvidedPem = utilities.convertDERtoPEM(deviceKeyBuffer);
+		const deviceID = deviceIDBuffer.toString('hex');
 
+		// todo remove stages;
     this._handshakeStage = 'read-core-id';
 
-    return deviceProvidedPem;
+    return { deviceID, deviceProvidedPem };
   };
 
-  // 4.) Read the public key from disk for this core
-  // TODO do this with keys repository?
-  _getDeviceKey = (deviceProvidedPem: string): Object => {
-    const publicKey = utilities.get_core_key(this._deviceID);
-    try {
-      if (!publicKey) {
-        this._handshakeFail(`couldn't find key for device: ${this._deviceID}`);
-        if (deviceProvidedPem) {
-          utilities.save_handshake_key(this._deviceID, deviceProvidedPem);
-        }
-        throw `Failed finding key for core: ${this._deviceID}`;
+
+  _getDevicePublicKey = async (
+    deviceID: string,
+    deviceProvidedPem: ?string,
+  ): Promise<Object> => {
+    const publicKeyString = await this._deviceKeyRepository.getById(deviceID);
+
+    if (!publicKeyString) {
+      if (deviceProvidedPem) {
+        this._deviceKeyRepository.update(deviceID, deviceProvidedPem);
+        return CryptoLib.createPublicKey(deviceProvidedPem);
       }
-    } catch (exception) {
-      logger.error('Error handling get_corekey ', exception);
-      this._handshakeFail(
-        `Failed handling find key for core: ${this._deviceID}`,
-      );
+
+      throw new Error(`no public key found for device: ${deviceID}`);
     }
 
     this._handshakeStage = 'get-core-key';
-    return publicKey;
+    return CryptoLib.createPublicKey(publicKeyString);
```

Comment:
I can do this ` CryptoLib.createPublicKey(publicKeyString);` inside _deviceKeyRepository.getById(), would it be better?